src/graph.rs

```diff
@@ -332,12 +529,90 @@ impl ModuleGraph {
     specifier: &ModuleSpecifier,
   ) -> Result<Option<&Module>, ModuleGraphError> {
     let specifier = self.resolve(specifier);
-    match self.modules.get(&specifier) {
+    match self.module_slots.get(&specifier) {
       Some(ModuleSlot::Module(module)) => Ok(Some(module)),
       Some(ModuleSlot::Err(err)) => Err(err.clone()),
       _ => Ok(None),
     }
   }
+
+  /// Walk the graph from the root, checking to see if there are any module
+  /// graph errors on non-dynamic imports. The first error is returned as an
+  /// error result, otherwise ok if there are no errors.
+  #[cfg(feature = "rust")]
+  pub fn valid(&self) -> Result<(), (ModuleSpecifier, ModuleGraphError)> {
+    fn validate<F>(
+      specifier: &ModuleSpecifier,
+      seen: &mut HashSet<ModuleSpecifier>,
+      get_module: &F,
+    ) -> Result<(), (ModuleSpecifier, ModuleGraphError)>
+    where
+      F: Fn(&ModuleSpecifier) -> Result<Option<Module>, ModuleGraphError>,
+    {
+      if seen.contains(specifier) {
+        return Ok(());
+      }
+      seen.insert(specifier.clone());
+      match get_module(specifier) {
+        Ok(Some(module)) => {
+          if let Some((_, Resolved::Specifier(specifier, _))) =
+            &module.maybe_types_dependency
+          {
+            validate(specifier, seen, get_module)?;
+          }
+          for dep in module.dependencies.values() {
+            if !dep.is_dynamic {
+              match &dep.maybe_code {
+                Resolved::Specifier(specifier, _) => {
+                  validate(specifier, seen, get_module)?
+                }
+                Resolved::Err(err, span) => {
```

Comment:
It might be nice to store the `span` in `ResolutionError` or do something so that we can change `maybe_code` to a `Result` and we could just do `dep.maybe_code?` above to return the error. That would also help reduce duplication here mentioned in my comment above because we'd only after to do one if let statement that calls `validate` in the body for both `maybe_code` and `maybe_type`.

Reply:
Yeah, agreed.